modules/ximgproc/src/scansegment.cpp

```diff
@@ -0,0 +1,926 @@
+			vertDiv = width > height ? shortCount : longCount;
+			horzLength = (float)width / (float)horzDiv;
+			vertLength = (float)height / (float)vertDiv;
+			effectivethreads = horzDiv * vertDiv;
+			smallClusters = 0;
+
+			// get array of seed rects
+			seedRects = static_cast<cv::Rect*>(malloc(horzDiv * vertDiv * sizeof(cv::Rect)));
+			seedRectsExt = static_cast<cv::Rect*>(malloc(horzDiv * vertDiv * sizeof(cv::Rect)));
+			offsetRects = static_cast<cv::Rect*>(malloc(horzDiv * vertDiv * sizeof(cv::Rect)));
+			for (int y = 0; y < vertDiv; y++) {
+				for (int x = 0; x < horzDiv; x++) {
+					int xStart = (int)((float)x * horzLength);
+					int yStart = (int)((float)y * vertLength);
+					cv::Rect seedRect = cv::Rect(xStart, yStart, (int)(x == horzDiv - 1 ? width - xStart : horzLength), (int)(y == vertDiv - 1 ? height - yStart : vertLength));
+
+					int bnd_l = seedRect.x;
+					int bnd_t = seedRect.y;
+					int bnd_r = seedRect.x + seedRect.width - 1;
+					int bnd_b = seedRect.y + seedRect.height - 1;
+					if (bnd_l > 0) {
+						bnd_l -= 1;
+					}
+					if (bnd_t > 0) {
+						bnd_t -= 1;
+					}
+					if (bnd_r < width - 1) {
+						bnd_r += 1;
+					}
+					if (bnd_b < height - 1) {
+						bnd_b += 1;
+					}
+
+					seedRects[(y * horzDiv) + x] = seedRect;
+					seedRectsExt[(y * horzDiv) + x] = cv::Rect(bnd_l, bnd_t, bnd_r - bnd_l + 1, bnd_b - bnd_t + 1);
+					offsetRects[(y * horzDiv) + x] = cv::Rect(seedRect.x - bnd_l, seedRect.y - bnd_t, seedRect.width, seedRect.height);
+				}
+			}
+
+			// get adjusted tolerance = (100 / average length (horz/vert)) x sqrt(3) [ie. euclidean lab colour distance sqrt(l2 + a2 + b2)] x tolerance100
+			adjTolerance = (200.0f / (width + height)) * sqrtf(3) * tolerance100;
+			adjTolerance = adjTolerance * adjTolerance;
+
+			// create neighbour vector
+			indexNeighbourVec = std::vector<int>(effectivethreads);
+			std::iota(indexNeighbourVec.begin(), indexNeighbourVec.end(), 0);
+
+			// create process vector
+			indexProcessVec = std::vector<std::pair<int, int>>(processthreads);
+			int processDiv = indexSize / processthreads;
+			int processCurrent = 0;
+			for (int i = 0; i < processthreads - 1; i++) {
+				indexProcessVec[i] = std::make_pair(processCurrent, processCurrent + processDiv);
+				processCurrent += processDiv;
+			}
+			indexProcessVec[processthreads - 1] = std::make_pair(processCurrent, indexSize);
+
+			// create buffers and initialise
+			std::vector<cv::Point> tempLoc{ cv::Point(-1, -1), cv::Point(0, -1), cv::Point(1, -1), cv::Point(-1, 0), cv::Point(1, 0), cv::Point(-1, 1), cv::Point(0, 1), cv::Point(1, 1) };
+			neighbourLoc = static_cast<cv::Point*>(malloc(8 * sizeof(cv::Point)));
+			memcpy(neighbourLoc, tempLoc.data(), 8 * sizeof(cv::Point));
+
+			labelsBuffer = static_cast<int*>(malloc(indexSize * sizeof(int)));
+			clusterBuffer = static_cast<int*>(malloc(indexSize * sizeof(int)));
+			pixelBuffer = static_cast<uchar*>(malloc(indexSize));
```

Comment:
Manual memory management should be avoided (malloc/new).
Use `std::vector` or `cv::AutoBuffer` instead.

Reply:
I did a test of malloc vs vector performance previously, and manual memory management is much more efficient.

Ran the following short test with a buffer size of 100000 over 100000 iterations. Results were malloc 1477646 us, vector 2696077 us, which is 45.2% faster.

void testMalloc(int iterate, int buffersize, int* malloctime, int* vectortime)
{
    auto tstart1 = std::chrono::high_resolution_clock::now();

    for (int i = 0; i < iterate; i++) {
        int* testBuffer = static_cast<int*>(malloc(buffersize * sizeof(int)));
        std::fill(testBuffer, testBuffer + buffersize, 0);
        free(testBuffer);
    }

    auto tend1 = std::chrono::high_resolution_clock::now();
    *malloctime = (int)std::chrono::duration_cast<std::chrono::microseconds>(tend1 - tstart1).count();

    auto tstart2 = std::chrono::high_resolution_clock::now();

    for (int i = 0; i < iterate; i++) {
        std::vector<int> testVector(buffersize);
        std::fill(testVector.begin(), testVector.end(), 0);
        testVector.resize(0);
    }

    auto tend2 = std::chrono::high_resolution_clock::now();
    *vectortime = (int)std::chrono::duration_cast<std::chrono::microseconds>(tend2 - tstart2).count();
}

Reply:
>    std::vector<int> testVector(buffersize);
>    std::fill(testVector.begin(), testVector.end(), 0);

`std::vector` already initializes (with zeros) in constructor.
Your micro benchmark dooes this twice.

There is `cv::AutoBuffer<int, 0>` which doesn't initialize values on creation.

---

Manual memory management is not exception safe code.

Reply:
Thanks for the info. I used the autobuffer in the same way as in the CLAHE module, and the speed is the same as for malloc (approx). So I have changed the code to reflect this.

Reply:
Hi, could I check if the removal of malloc and replacement with autobuffer is sufficient to ensure exception-safe code?